agent/grpc-external/services/resource/delete.go

```diff
@@ -0,0 +1,48 @@
+package resource
+
+import (
+	"context"
+	"errors"
+
+	"github.com/hashicorp/consul/internal/storage"
+	"github.com/hashicorp/consul/proto-public/pbresource"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+// Deletes the resource with the given Id and Version.
+// Pass an empty Version to delete a resource regardless of it's persisted Version.
+// Returns an error if the resource does not exist.
+// Returns an error if the passed in version doesn't match the persisted version.
+func (s *Server) Delete(ctx context.Context, req *pbresource.DeleteRequest) (*pbresource.DeleteResponse, error) {
+	// check type registered
+	reg, err := s.resolveType(req.Id.Type)
+	if err != nil {
+		return nil, err
+	}
+
+	// TODO(spatel): reg will be used for ACL hooks
+	_ = reg
+
+	if req.Version == "" {
+		// Delete resource regardless of the persisted Version. Hence, strong read
+		// necessary to get latest Version
+		existing, err := s.Backend.Read(ctx, storage.StrongConsistency, req.Id)
```

Comment:
FWIW, there's a chance that the resource will be updated between the `Read` and `DeleteCAS`, in which case the user would see `ErrCASFailure`.

It's pretty unlikely though, so I don't think we have to handle it with retries like we do in the `Write` endpoint. I guess it'd happen if a user was trying to delete a resource at the same time a controller was modifying it.

Reply:
👍 Fixed in https://github.com/hashicorp/consul/pull/16756/commits/d038a2538accb28fdad204653df3305e38058fda